File: frontend/js/screen.js
// screen.js — 퀀트 스크리닝 페이지 로직

function num(id) {
  const v = document.getElementById(id).value.trim();
  return v === "" ? null : Number(v);
}

function fmt(v, d = 2) {
  if (v === null || v === undefined) return "—";
  return Number(v).toLocaleString("ko-KR", { maximumFractionDigits: d });
}

function fmtCap(v, currency) {
  if (v === null || v === undefined) return "—";
  const unit = currency === "USD" ? 1e9 : 1e12;
  const label = currency === "USD" ? "B" : "조";
  return (v / unit).toLocaleString("ko-KR", { maximumFractionDigits: 1 }) + label;
}

function setStatus(msg, isError = false) {
  const el = document.getElementById("status");
  el.textContent = msg;
  el.className = "status" + (isError ? " error" : "");
  el.classList.remove("hidden");
}

function renderResults(data) {
  document.getElementById("status").classList.add("hidden");
  document.getElementById("summary").textContent =
    `전체 ${data.total}종목 중 조건 통과 ${data.passed}종목`;

  let html = `<table class="fin"><thead><tr>
    <th>결과</th><th>티커</th><th>종목명</th><th>현재가</th>
    <th>PER</th><th>PBR</th><th>ROE</th><th>배당%</th><th>시총</th><th>비고</th>
  </tr></thead><tbody>`;

  data.results.forEach((r) => {
    if (!r.ok) {
      html += `<tr><td class="r-fail">에러</td><td>${r.ticker}</td>
        <td colspan="8" class="reject">${r.error || "조회 실패"}</td></tr>`;
      return;
    }
    const badge = r.passed
      ? `<span class="r-pass">✔ 통과</span>`
      : `<span class="r-fail">✘ 탈락</span>`;
    const note = r.passed ? "" : r.reasons.join(", ");
    html += `<tr>
      <td class="${r.passed ? "r-pass" : "r-fail"}">${badge}</td>
      <td>${r.ticker}</td>
      <td style="text-align:left">${r.name || "—"}</td>
      <td>${fmt(r.price, r.currency === "KRW" ? 0 : 2)}</td>
      <td>${fmt(r.per)}</td>
      <td>${fmt(r.pbr)}</td>
      <td>${r.roe != null ? fmt(r.roe) + "%" : "—"}</td>
      <td>${r.dividend_yield != null ? fmt(r.dividend_yield) + "%" : "—"}</td>
      <td>${fmtCap(r.market_cap, r.currency)}</td>
      <td style="text-align:left" class="reject">${note}</td>
    </tr>`;
  });
  html += "</tbody></table>";
  document.getElementById("table").innerHTML = html;
  document.getElementById("result").classList.remove("hidden");
}

async function run() {
  const tickers = document
    .getElementById("tickers")
    .value.split(/[\s,]+/)
    .map((t) => t.trim())
    .filter(Boolean);

  if (!tickers.length) {
    setStatus("티커를 하나 이상 입력하세요.", true);
    return;
  }

  const filters = {
    per_max: num("per_max"),
    pbr_max: num("pbr_max"),
    roe_min: num("roe_min"),
    dividend_min: num("dividend_min"),
  };

  document.getElementById("result").classList.add("hidden");
  setStatus(`${tickers.length}종목 스크리닝 중... (종목이 많으면 수 초 소요)`);

  try {
    const data = await PioneerAPI.screen(tickers, filters);
    renderResults(data);
  } catch (e) {
    setStatus(`스크리닝 실패: ${e.message}`, true);
  }
}

document.getElementById("runBtn").addEventListener("click", run);

File: backend/routers/screen.py
"""
screen.py — 퀀트 스크리닝 라우터.
POST /api/screen
    { "tickers": ["AAPL","005930.KS",...],
      "filters": { "per_max": 20, "pbr_max": 2, "roe_min": 10,
                   "dividend_min": 1, "market_cap_min": 0 } }
→ 사용자가 입력한 티커들을 조건으로 필터링해 통과/탈락 결과를 반환.
"""
from pydantic import BaseModel, Field
from fastapi import APIRouter

from services import data_fetcher

router = APIRouter(prefix="/api", tags=["screen"])


class ScreenFilters(BaseModel):
    per_max: float | None = None          # PER 이하
    pbr_max: float | None = None          # PBR 이하
    roe_min: float | None = None          # ROE(%) 이상
    dividend_min: float | None = None     # 배당수익률(%) 이상
    market_cap_min: float | None = None   # 시가총액 이상(현지통화)


class ScreenRequest(BaseModel):
    tickers: list[str] = Field(..., min_length=1)
    filters: ScreenFilters = ScreenFilters()


def _check(info: dict, f: ScreenFilters) -> tuple[bool, list[str]]:
    """조건 통과 여부와 탈락 사유 목록을 반환한다. 값이 없으면 해당 조건은 통과 처리(불이익 없음)."""
    reasons: list[str] = []

    def fail(cond: bool, msg: str):
        if cond:
            reasons.append(msg)

    if f.per_max is not None and info.get("per") is not None:
        fail(info["per"] > f.per_max, f"PER {info['per']} > {f.per_max}")
    if f.pbr_max is not None and info.get("pbr") is not None:
        fail(info["pbr"] > f.pbr_max, f"PBR {info['pbr']} > {f.pbr_max}")
    if f.roe_min is not None and info.get("roe") is not None:
        fail(info["roe"] < f.roe_min, f"ROE {info['roe']} < {f.roe_min}")
    if f.dividend_min is not None and info.get("dividend_yield") is not None:
        fail(info["dividend_yield"] < f.dividend_min, f"배당 {info['dividend_yield']} < {f.dividend_min}")
    if f.market_cap_min is not None and info.get("market_cap") is not None:
        fail(info["market_cap"] < f.market_cap_min, f"시총 < {f.market_cap_min}")

    return (len(reasons) == 0, reasons)


@router.post("/screen")
def screen(req: ScreenRequest):
    results = []
    passed = 0
    # 중복 제거 + 정규화
    seen = set()
    for raw in req.tickers:
        ticker = raw.strip().upper()
        if not ticker or ticker in seen:
            continue
        seen.add(ticker)

        try:
            info = data_fetcher.get_stock_info(ticker)
        except Exception as e:
            results.append({
                "ticker": ticker, "ok": False, "error": str(e),
                "passed": False, "reasons": ["조회 실패"],
            })
            continue

        ok, reasons = _check(info, req.filters)
        if ok:
            passed += 1
        results.append({
            "ticker": ticker,
            "ok": True,
            "passed": ok,
            "reasons": reasons,
            "name": info.get("name"),
            "currency": info.get("currency"),
            "price": info.get("price"),
            "per": info.get("per"),
            "pbr": info.get("pbr"),
            "roe": info.get("roe"),
            "dividend_yield": info.get("dividend_yield"),
            "market_cap": info.get("market_cap"),
        })

    # 통과 종목을 위로 정렬
    results.sort(key=lambda r: (not r["passed"], r["ticker"]))
    return {
        "total": len(results),
        "passed": passed,
        "filters": req.filters.model_dump(),
        "results": results,
    }
